multi-channel-chat/nats-channel/src/lib.rs

```diff
@@ -0,0 +1,123 @@
+
+#[async_trait]
+impl Outbound for NatsChannelActor {
+    async fn publish_message(&self, ctx: &Context, arg: &OutboundMessage) -> RpcResult<bool> {
+        let pubber = MessagingSender::new();
+        let res = pubber
+            .publish(
+                ctx,
+                &PubMessage {
+                    subject: RTCHAT_TOPIC.to_string(),
+                    reply_to: None,
+                    body: serde_json::to_vec(arg).unwrap(),
+                },
+            )
+            .await;
+        Ok(res.is_ok())
+    }
+}
+
+/// Implementation of Chat Log actor trait methods
+#[async_trait]
+impl MessageSubscriber for NatsChannelActor {
+    async fn handle_message(&self, ctx: &Context, msg: &SubMessage) -> RpcResult<()> {
+        let reply_topic = msg
+            .reply_to
+            .as_ref()
+            .unwrap_or(&"dedletter".to_string())
+            .to_string();
+        match serde_json::from_slice::<IncomingMessage>(&msg.body) {
+            Ok(im) => {
+                let logger = ChatlogSender::to_actor(API_ACTOR);
+                let numgen = NumberGenSender::new();
+                let guid = numgen.generate_guid(ctx).await.unwrap_or("n/a".to_string());
+                match logger
+                    .write_message(
+                        ctx,
+                        &CanonicalChatMessage {
+                            body: im.body,
+                            channel_name: CHANNEL_NAME.to_string(),
+                            id: guid,
+                            source_user: im.user_name,
+                        },
+                    )
+                    .await
+                {
+                    Ok(resp) => {
+                        let ack = Ack {
+                            accepted: resp.accepted,
+                            reason: resp.reason,
+                        };
+                        let _ = pub_ack(ctx, &reply_topic, ack).await;
+                    }
+                    Err(e) => {
+                        let _ = pub_fail(ctx, &reply_topic, format!("{}", e)).await;
+                    }
+                }
+            }
+            Err(_) => {
+                let _ = pub_fail(ctx, &reply_topic, "Bad Request".to_string()).await;
+            }
+        }
+        Ok(())
+    }
+}
+
+async fn pub_ack(ctx: &Context, topic: &str, ack: Ack) {
+    let publisher = MessagingSender::new();
+    let _ = publisher
+        .publish(
+            ctx,
+            &PubMessage {
+                subject: topic.to_string(),
+                reply_to: None,
+                body: serde_json::to_vec(&ack).unwrap(),
+            },
+        )
+        .await;
+}
+
+async fn pub_fail(ctx: &Context, topic: &str, msg: String) {
+    pub_ack(
+        ctx,
+        topic,
+        Ack {
+            accepted: false,
+            reason: Some(msg),
+        },
+    )
+    .await
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct IncomingMessage {
```

Comment:
Super duper nitty nit: This could use `&str` for its data members and be a zero copy deserialization. Definitely not required here

```rust
struct IncomingMessage<'a> {
    user_name: &'a str,
    body: &'a str,
}
```

Note that you could do the same thing with the ack to avoid cloning data

Reply:
IMHO this syntax looks really really repellant to non-Rust people. if we're trying to be easy to read then I'd like to leave lifetimes out of this.